### DISCUSSION TITLE
Option to disable "smart" watcher

### DISCUSSION THREAD

**Discussion by cdimitroulas** [avajs/ava#2887] [Ideas]
I often find that the "smart" watcher gets in my way and I'd rather just rerun all the tests anytime any file is changed.

For example, when writing integration tests Ava is often not able to determine that a file change actually affects a particular test and so then it decides not to run that test again. This results in me having to go to that terminal and hit "r + Enter" in order to manually tell Ava to rerun all the tests.

It would be great if there was a flag or option which would disable this "smart" behaviour of Ava's `--watch` mode. The result would be that any file change would trigger a full rerun of all the tests.

**Comment by novemberborn** [comment]
AVA is supposed to re-run all tests if it can't attribute a changed file to a single test. Is the problem here that it's only partially attributing the changes, or is there something else going on?

Could you run watch mode with `DEBUG=ava:* npx ava --watch` and share the output? That should show you the changes and the decision making process.

I assume this is with AVA 3? There are fixes in AVA 4 (pre-release) but I think those fix accidental re-runs of all tests, so that won't help here.

**Reply by cdimitroulas** [reply]
Yes, this is with AVA 3. I'll try running it with the debug command to see if I can determine what causes this issue for me, thanks

**Reply by cdimitroulas** [reply]
It seems that AVA is determining a dependency, which is the `app.ts` file (because I'm importing the whole app to run integration tests against it) and then only running a couple of tests which are actually in an `app.test.ts` file. Those `app.test.ts` test cases are just some very basic high-level tests for the app, but most of the integration tests reside in other files which also import from `app.ts`.

It seems that AVA doesn't realize that those other tests also depend on `app.ts` for some reason.

In any case, I think there will always be cases where the decision making process might not be clever enough to figure out what you want, so having an option to disable it and force a full re-run on each every code change would be useful, at least for myself :)

**Reply by novemberborn** [reply]
Would you be able to share the debug output and some more details about your setup?

**Reply by cdimitroulas** [reply]
It's a fairly common Koa application (similar to Express, it's basically the same middleware-style framework). I won't share any of the code as it's from my work and I am not able to do so, but I can outline a very simplified version of what our setup looks like.

Routes are defined in files which export a router that can be mounted on the app e.g.
```ts
// users.ts
import { Router } from 'koa'

export const userRouter = new Router()

userRouter.get("/", (ctx) => { /* ... */ })

userRouter.post("/", (ctx) => { /* ... */ })

// etc.
```

The app.ts file looks something like
```ts
import Koa from 'koa'
import { userRouter } from './users'

export const app = new Koa()
app.use(bodyParser())
app.use(userRouter.routes())
// etc.
```

The test file simply imports the app from `app.ts` like so
```ts
// mytest.test.ts
import { app } from '../app'
```

The problem is that when writing integration tests which test the app by calling endpoints defined in `users.ts`, changes to `users.ts` don't cause those tests to rerun. It seems that the dependency tracking logic in AVA 3 isn't able to figure it out properly.

The output in my particular case looks like this:
```
  ava:watcher Detected change of src/modules/subscriptions/index.ts +0ms
  ava:watcher /<redacted>/src/modules/subscriptions/index.ts is a dependency of /<redacted>/src/app.test
.ts +101ms
```

**Reply by novemberborn** [reply]
Are you using full CommonJS or is there a mix of ESM and CJS? There is no dependency tracking for ESM so that may be an issue if there's a mix.

How are you handling TypeScript? Maybe that's hooked up in such a way that AVA can only track partial dependencies.

**Reply by cdimitroulas** [reply]
We are not using ESM, only CommonJS. 

This is our ava.config.js file:
```js
export default {
  extensions: ["ts", "tsx"],
  files: ["src/**/*.test.{js,jsx,ts,tsx}"],
  require: ["ts-node/register/transpile-only"],
};
```

**Reply by novemberborn** [reply]
@cdimitroulas I was going to turn this into a feature request, but then I noticed something. Did the watcher observe the change to `users.ts`? Why did `index.ts` change?

**Reply by cdimitroulas** [reply]
The `users.ts` comment in the original code was just an example, it doesn't represent a real file in our app.

The real file was the `subscriptions/index.ts` that you can see detected in the watcher output. The watcher only understands that this is a dependency of `/<redacted>/src/app.test.ts` but doesn't understand that it's also a dependency of `/<redacted>/src/modules/__tests__/someOtherTest.test.ts` where the `app` is also imported

**Reply by novemberborn** [reply]
Sorry for being slow with this, both in ability to respond as figuring out the underlying problem.

I've tried to reproduce this with AVA 4, using straight-forward source and test files and `ts-node/register/transpile-only`. It is all working as expected.

As you've pointed out the problem is in the dependency tracking. It's not an issue we've come across but then the dependency tracking is a creative hack of the CommonJS module system so it shouldn't be too surprising that there's a poorly understood failure mode.

I've opened https://github.com/avajs/ava/issues/2905 to have an option to disable dependency tracking.

**Reply by cdimitroulas** [reply]
Thank you!
